Dockerfile

```diff
@@ -103,13 +110,47 @@ RUN --mount=target=. --mount=target=/root/.cache,type=cache \
   xx-go build ${GOBUILDFLAGS} -ldflags "$(cat /tmp/.ldflags) -extldflags '-static'" -tags "osusergo netgo static_build seccomp ${BUILDKITD_TAGS}" -o /usr/bin/buildkitd ./cmd/buildkitd && \
   xx-verify ${VERIFYFLAGS} /usr/bin/buildkitd
 
-FROM scratch AS binaries-linux
+# build buildkitd binary (debug variant)
+FROM buildkit-base AS buildkitd-debug
+# BUILDKITD_TAGS defines additional Go build tags for compiling buildkitd
+ARG BUILDKITD_TAGS
+ARG TARGETPLATFORM
+ARG GOBUILDFLAGS
+ARG VERIFYFLAGS="--static"
+ARG CGO_ENABLED=0
+RUN --mount=target=. --mount=target=/root/.cache,type=cache \
+  --mount=target=/go/pkg/mod,type=cache \
+  --mount=source=/tmp/.ldflags,target=/tmp/.ldflags,from=buildkit-version \
+  xx-go build ${GOBUILDFLAGS} -gcflags="all=-N -l" -ldflags "$(cat /tmp/.ldflags) -extldflags '-static'" -tags "osusergo netgo static_build seccomp ${BUILDKITD_TAGS}" -o /usr/bin/buildkitd ./cmd/buildkitd && \
+  xx-verify ${VERIFYFLAGS} /usr/bin/buildkitd
+
+FROM buildkitd-${BUILD_VARIANT} AS buildkitd
+
+FROM scratch AS binaries-linux-base
 COPY --link --from=runc /usr/bin/runc /buildkit-runc
 # built from https://github.com/tonistiigi/binfmt/releases/tag/buildkit%2Fv7.1.0-30
 COPY --link --from=tonistiigi/binfmt:buildkit-v7.1.0-30@sha256:45dd57b4ba2f24e2354f71f1e4e51f073cb7a28fd848ce6f5f2a7701142a6bf0 / /
 COPY --link --from=buildctl /usr/bin/buildctl /
+
+FROM binaries-linux-base AS binaries-linux-release
 COPY --link --from=buildkitd /usr/bin/buildkitd /
 
+FROM binaries-linux-base AS binaries-linux-debug
+COPY --link --from=buildkitd /usr/bin/buildkitd /buildkitd.bin
+COPY --link --from=dlv /usr/bin/dlv /
+COPY --link --chmod=755 <<EOF /buildkitd
+#!/bin/sh
+exec dlv exec /usr/bin/buildkitd.bin \\
+  --api-version=2 \\
+  -l 0.0.0.0:5000 \\
+  --headless=true \\
+  --accept-multiclient \\
```

Comment:
Is this the only config that makes sense for buildkitd?

Reply:
Not necessarily and this is a bit of an unknown for me. I'm mostly familiar with how to connect this by the command line or the JetBrains Go Remote plugin. I'll go over each parameter.

* `--api-version=2` - I think this exists for backwards compatibility. I don't know the extent, but they seem to always suggest running this with `--api-version=2` and the Go Remote task only supports the version 2 API. I believe that you can still connect to version 1 even when version 2 is used.
* `-l 0.0.0.0:5000` - You can change the port and that's fine, but using only one port isn't a big deal since you can freely rebind ports for the host machine anyway. I don't know if unix sockets can be used, but I haven't seen that and the Go Remote plugin doesn't provide that option anyway. The any interface must be used because otherwise docker won't be able to bind to it.
* `--headless=true` - I think this is the only one that makes sense.
* `--accept-multiclient` - You could feasibly remove this, but there's also no downside to using it.
* `--continue` - This one could feasibly be different. If this is not set, then the program won't start until the debugger connects which means that it won't start listening on the port. It's a bit of a choice whether to include this or not. I thought this would be better because then you can use the debug image without necessarily using a debugger. But, if you're debugging the startup of buildkit, you'd want to remove this flag.